File: backend/app/pipeline.py
from .services.powerbi import PowerBIDataExtractor
from .services.azure_storage import AzureDataProcessor
from .services.azure_search import AzureAIFoundryRAG
from .core.config import Settings
import time
from typing import Dict

class PowerBIToRAGPipeline:
    def __init__(self, settings: Settings):
        self.pbi_extractor = PowerBIDataExtractor(
            settings.CLIENT_ID,
            settings.CLIENT_SECRET,
            settings.TENANT_ID,
            settings.WORKSPACE_ID
        )

        self.azure_processor = AzureDataProcessor(
            settings.STORAGE_ACCOUNT_NAME,
            settings.STORAGE_ACCOUNT_KEY,
            settings.STORAGE_CONTAINER
        )

        self.rag_system = AzureAIFoundryRAG(
            settings.SEARCH_SERVICE_NAME,
            settings.SEARCH_ADMIN_KEY,
            settings.SEARCH_INDEX_NAME
        )

    def run_pipeline(self, dataset_id: str, dax_queries: Dict[str, str]):
        """Execute complete pipeline from Power BI to RAG system"""
        print("Starting Power BI to RAG Pipeline...")

        extracted_data = {}
        processed_dfs = {}

        for query_name, dax_query in dax_queries.items():
            print(f"   Executing query: {query_name}")
            result = self.pbi_extractor.execute_dax_query(dataset_id, dax_query)

            if result:
                df = self.azure_processor.process_powerbi_response(result)
                if df is not None:
                    extracted_data[query_name] = result
                    processed_dfs[query_name] = df

                    blob_name = f"powerbi_data/{query_name}_{int(time.time())}.parquet"
                    self.azure_processor.save_to_parquet(df, blob_name)
                    print(f"   Saved {query_name} to blob storage")

        print("2. Indexing data for RAG system...")
        if processed_dfs:
            success = self.rag_system.index_powerbi_data(processed_dfs)
            if success:
                print("   Successfully indexed data in Azure AI Search")
            else:
                print("   Failed to index data")

        return {
            "extracted_data": extracted_data,
            "processed_dataframes": processed_dfs,
            "pipeline_status": "completed" if processed_dfs else "failed"
        }

File: backend/src/services/pipeline.js
// Import the services that the pipeline will orchestrate
import PowerBIService from './powerbi.js';
import AzureStorageService from './azureStorage.js';
import AzureSearchService from './azureSearch.js';

/**
 * @class PipelineService
 * @description Orchestrates the entire data extraction and indexing pipeline.
 * This class coordinates the process of fetching data from Power BI,
 * processing it, storing it in Azure Blob Storage, and finally indexing it
 * in Azure AI Search.
 */
class PipelineService {
  constructor() {
    // Instantiate the services that the pipeline will use
    this.powerBIService = new PowerBIService();
    this.azureStorageService = new AzureStorageService();
    this.azureSearchService = new AzureSearchService();
  }

  /**
   * @method run
   * @description Executes the complete data pipeline.
   * This method takes a dataset ID and a set of DAX queries, then runs the
   * full sequence of operations to get the data into the search index.
   * @param {string} datasetId - The ID of the Power BI dataset to query.
   * @param {Object.<string, string>} daxQueries - A dictionary of DAX queries to execute.
   * @returns {Promise<object>} An object containing the results of the pipeline execution.
   */
  async run(datasetId, daxQueries) {
    console.log('Starting Power BI to RAG Pipeline...');
    const extractedData = {};
    const processedData = {};

    // First, create the search index if it doesn't exist
    await this.azureSearchService.createSearchIndex();

    // Iterate over each DAX query and process it
    for (const [queryName, daxQuery] of Object.entries(daxQueries)) {
      console.log(`   Executing query: ${queryName}`);
      // Execute the DAX query using the Power BI service
      const result = await this.powerBIService.executeDaxQuery(datasetId, daxQuery);

      if (result) {
        // Process the response to get a structured format
        const processed = this.azureStorageService.processPowerBiResponse(result);
        if (processed) {
          extractedData[queryName] = result;
          processedData[queryName] = processed;

          // Save the processed data to a Parquet file in Azure Blob Storage
          const blobName = `powerbi_data/${queryName}_${Date.now()}.parquet`;
          await this.azureStorageService.saveToParquet(processed, blobName);
          console.log(`   Saved ${queryName} to blob storage`);
        }
      }
    }

    // If there is processed data, index it in Azure AI Search
    if (Object.keys(processedData).length > 0) {
      console.log('Indexing data for RAG system...');
      const success = await this.azureSearchService.indexPowerBIData(processedData);
      if (success) {
        console.log('   Successfully indexed data in Azure AI Search');
      } else {
        console.log('   Failed to index data');
      }
    }

    // Return a summary of the pipeline's execution
    return {
      extractedData: extractedData,
      processedData: processedData,
      pipelineStatus: Object.keys(processedData).length > 0 ? 'completed' : 'failed',
    };
  }
}

// Export the PipelineService class
export default PipelineService;
